Fix recursive scanning of directories with cyclic symlinks

uklotzde opened the pull request:
Just a quick fix that provides a workaround for the cyclic symlink scanning bug:

https://bugs.launchpad.net/mixxx/+bug/1493594

Neither elegant nor exhaustive (see my comments in the bug description), but it works.

uklotzde committed:
Fix recursive scanning of directories with cyclic symlinks

uklotzde committed:
Avoid to scan the top-level directory twice in presence of cyclic links

ywwg commented on src/library/scanner/recursivescandirectorytask.cpp:
```diff
@@ -59,12 +63,21 @@ void RecursiveScanDirectoryTask::run() {
+            const QDir currentDir(currentFile);
+            const QString canonicalDirPath(currentDir.canonicalPath());
+            if (m_scannerGlobal->isDirectoryScanned(canonicalDirPath)) {
+                // Skip directories that have already been scanned
```
your code is so straightforward I think this comment is not necessary

ywwg commented on src/library/scanner/recursivescandirectorytask.cpp:
```diff
@@ -59,12 +63,21 @@ void RecursiveScanDirectoryTask::run() {
+            }
+            // Add unvisited directories to our list of directories to scan.
+            dirsToScan.append(currentDir);
             // Skip the iTunes Album Art Folder since it is probably a waste of
```
this comment should be deleted or rephrased and moved above to the Blacklisted check

uklotzde replied:
I'll leave this comment there, just to be consistent.

On 09/14/2015 03:07 AM, Owen Williams wrote:
>
> In src/library/scanner/recursivescandirectorytask.cpp 
> <https://github.com/mixxxdj/mixxx/pull/708#discussion_r39355453>:
>
> > @@ -59,12 +63,21 @@ void RecursiveScanDirectoryTask::run() {
> >                  possibleCovers.append(currentFileInfo);
> >              }
> >          } else {
> > -            // File is a directory. Add it to our list of directories to scan.
> > +            // File is a directory
> > +            if (m_scannerGlobal->directoryBlacklisted(currentFile)) {
> > +                // Skip blacklisted directories
> > +                continue;
> > +            }
> > +            const QDir currentDir(currentFile);
> > +            const QString canonicalDirPath(currentDir.canonicalPath());
> > +            if (m_scannerGlobal->isDirectoryScanned(canonicalDirPath)) {
> > +                // Skip directories that have already been scanned
> > +                continue;
> > +            }
> > +            // Add unvisited directories to our list of directories to scan.
> > +            dirsToScan.append(currentDir);
> >              // Skip the iTunes Album Art Folder since it is probably a waste of
>
> this comment should be deleted or rephrased and moved above to the 
> Blacklisted check
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/708/files#r39355453>.
>

uklotzde committed:
Clarify some comments

ywwg commented on src/library/scanner/recursivescandirectorytask.cpp:
```diff
@@ -59,12 +63,20 @@ void RecursiveScanDirectoryTask::run() {
+            const QDir currentDir(currentFile);
+            const QString canonicalDirPath(currentDir.canonicalPath());
+            if (m_scannerGlobal->isDirectoryScanned(canonicalDirPath)) {
+                // Skip directories that have already been scanned.
```
like I said though, you don't need this comment

ywwg commented on src/library/scanner/recursivescandirectorytask.cpp:
```diff
@@ -59,12 +63,20 @@ void RecursiveScanDirectoryTask::run() {
+                // Skip directories that have already been scanned.
+                continue;
+            }
+            // Add unvisited directories to our list of directories to scan.
```
or this one.  "dirstoscan.append" is very clear already

ywwg commented:
what sorts of library tests do we have?  Is this something we could write a test for?

kain88-de commented:
@ywwg It should be possible to create a temprary file structure. The QFile object has a [method](http://doc.qt.io/qt-4.8/qfile.html#link) to create symbolic links. Afterwards we can check if the library state is what we expect.

The only thing I'm not sure right now is if we can start the scanner in the unit-tests.

ywwg commented:
It might be hard to make a multi-platform test for this, although I wouldn't mind some stupid ifdefs if it helps.  We don't necessarily have to write a test for this fix, but if not, I'd like @uklotzde to do some detailed hand-testing and report back.  For example, make a test library root, make some symlink loops, do some tricky things, etc.  Doesn't have to be more than 5 files to scan, what's more interesting is the directory structure.

ywwg commented:
Alternatively, if there's anyone out there who is not comfortable with developing C++ but is comfortable with github and compiling, they could do some QA and really help us out!

uklotzde commented:
I have only performed some manual testing with some weird cyclic links 
including the top-level directory of the library. But I think about creating 
a unit test for this.

There is one more thing that I would like to do: Postpone scanning of 
directories that are symbolic links until all regular directories have been 
scanned. This strategy should result in reasonable track locations with a 
minimum amount of symbolic links, even if it is no strict ordering.

On 09/14/2015 04:16 PM, Owen Williams wrote:
>
> It might be hard to make a multi-platform test for this, although I 
> wouldn't mind some stupid ifdefs if it helps. We don't necessarily have to 
> write a test for this fix, but if not, I'd like @uklotzde 
> <https://github.com/uklotzde> to do some detailed hand-testing and report 
> back. For example, make a test library root, make some symlink loops, do 
> some tricky things, etc. Doesn't have to be more than 5 files to scan, 
> what's more interesting is the directory structure.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/708#issuecomment-140092427>.
>

ywwg commented:
> Postpone scanning of
> directories that are symbolic links until all regular directories have been
> scanned.

I don't really see a compelling reason to do that... I'd be more interested in fixing the removable-media use-case which is a much bigger problem for users.

Thanks for looking in to the possibility of a unit test.

daschuer commented:
I have tested this with various links and it works. 
However, there is a race condition writing and reading the scanned directories.
We need a locked function, that checks and stores the canonical path at once.

I agree with Uwe that there is additional work required to ensure that a scan will always have the same results, using the same directory structure. Currently we cannot guarantee that due to the  concurrently nature of the scanner threads. 
However, we have to discuss the strategy, that on one hand has a reliable repeatable result, but has no performance penalty. Delaying the links will probably not work for my library where I have links to my removable drives in my Music directory.

daschuer commented:
Maybe it it works to have a pre-scan stage, that checks if we have scanned a directory by an alternative path and is still accessible by the previous path. If yes, it should keep the previous path, currently Mixxx assumes the tracks where moved.

uklotzde committed:
Delete unnecessary code comments (review)

uklotzde committed:
Avoid race condition while scanning directories (review)

ywwg commented:
> Currently we cannot guarantee that due to the concurrently nature of the scanner threads.

what does this mean, that every time I start up mixxx files in symlinked folders might move around from one apparent folder to another?

daschuer commented:
> what does this mean, ...

If you have two symlinks to the same folder, they are scanned by two separate threads. It depends on the scheduler and disc load, when they are scheduled. If the order changes between two scans, Mixxx will assume tracks have moved.

daschuer commented:
LGTM, Thank you very much Uwe!

uklotzde commented:
Unit testing proves to be difficult, because the code has lots of 
dependencies. Even if you don't follow TDD ruthlessly, thinking about how to 
test the code upfront usually leads to a better, cleaner design.

I'm currently trying to tame the TIO beast and don't want to open up another 
pandora's box. Of course, I'm trying to write unit tests for each new class 
that I introduce in this context ;)

On 09/14/2015 05:47 PM, Owen Williams wrote:
>
>     Postpone scanning of
>     directories that are symbolic links until all regular directories have
>     been
>     scanned.
>
> I don't really see a compelling reason to do that... I'd be more 
> interested in fixing the removable-media use-case which is a much bigger 
> problem for users.
>
> Thanks for looking in to the possibility of a unit test.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/708#issuecomment-140121464>.
>

ywwg commented:
> Even if you don't follow TDD ruthlessly,

Mixxx is a very old codebase, so I don't hold the lack of testability against it.  We should all buy copies of "Working Effectively With Legacy Code" for strategies to help in these situations.



> If the order changes between two scans, Mixxx will assume tracks have moved.

Hm, I don't think this is good.  If a user has a lot of symlinks then a lot of tracks are going to move around.  Is this something that is a problem now, or just with this new code?  I don't want to merge until I have a better sense of that bug.

daschuer commented:
> Is this something that is a problem now, or just with this new code?

* Old code: Track is listed twice

* New code, fresh DB: Track may move

* Updaters: On of the duplicated tracks is marked as missing. Which it is may change between scans. 

* Downgraders: duplicated track is back.  

.. but this is a different issue form the original one posted in the bug report. This PR actually fixes one issue and does not make the other issue worse nor messes unrevertable with the users library.

ywwg commented:
I do think this change makes things worse.  Tracks marked as missing are removed from the library view, so if the user has always loaded the track that mixxx decides is redundant, they will lose their tagging.  They might even have different tags depending on which symlinked version of the file they are using (I could imagine someone using symlinks to build a playlist.) 

We need to be able to solve the issue of symlink cycles without breaking the current behavior.

daschuer commented:
I do not think it is a real use case that one tracks two sets of metadata for a single physical file. 

On the other hand we have users complaining about duplicated tracks in their library. 

So I consider it as an undesired situation. 

I think we will fix the remaining issue by removing all duplicates from the library, and merge the metadata somehow. So hiding them after merging this PR is on the track. 
The missing tracks are still part of the playlists / crates and are still playable.

ywwg commented:
I use the Location column of the library extensively, it would be very weird to have tracks changing location on every startup.  We need to fix that.

daschuer commented:
Do you have duplicated Tracks? What is your use case for this? 
If not you are like the majority of users not effected. 

You will have the issue that the played flag and played counter will only show up at one instance of the track. 

Fixing this can't be done in a quick hack, so lets merge this and postpone a solution for duplicates to 
the time we have a rock solid solution for duplicates.

daschuer commented:
By the way: do we really rescan the library on every start-up?

ywwg commented:
> every start-up?

It's a preference option.  I'm not sure what the default is.

ywwg commented:
OK so from what I understand, the first time someone runs this code, if they have duplicate tracks, one of the copies will disappear and be marked as missing.  Let's say the user edits the comment field.  When the user restarts, if mixxx finds the other track first, what happens?  Is the track considered moved, or are the two tracks exchanged?  Like:

Track A:   comment is "foo"
Track A' (symlink): comment is "bar"

rescan... it finds A.  A' is missing

rescan... it finds A'.  does it consider A->A', so the comment is "foo", or does it consider A missing and A' found, so the comment is "bar"?

daschuer commented:
Track A: comment is "foo"
Track A' (symlink): comment is "bar"

rescan... it finds A. A' is missing
rescan... it finds A'. A is missing

It does not consider that a track is moved, because Mixxx has not found a new Track. 
The Track in shown in the library cahnges from A to A' and the comment accordingly "foo" -> "bar"

The As and A's in the playlist and crates remains untouched.   

I expect that such a swap on a normal restart to be a rare case since it requires a directory hash change the track and the link to be on two scanning routes on nearly equal time. It is more likely that a swap happens if the user has changes pats of the library tree.

ywwg commented:
ugh, so that means every time the user restarts they could see the comment swap back and forth between "foo" and "bar".  that is not acceptable :(.

daschuer commented:
They could see a swap under some rare circumstances. 
So we are talking about a rare issue for the minority of users. 
Are you among them? 

I consider a duplicated track by a symlink as a not supported feature in 1.11.
Even in 1.11 the user will see the swap if he has one instance in one crate and the other instance in the other carte.  

This PR helps that no new user runs into this condition, which is highly desirable for me. 
A fresh scan only suffers the issue that the path may changes, but the play-count is tracked correctly.

ywwg commented:
I don't use symlinks.  I think it's acceptable to get into a state where the location change between restarts, but we need a migration path for users with existing duplicate track entries.  I think it'd be ok to just delete one of them from the database forcibly.  Is there a way we know which files share the same canonical path?

daschuer commented:
All "stupid" auto delete function makes the issue worth, since it may damage a crates and playlist.
Until we have a quite intelligent merge solution, I consider the remaining issues in this PR as rare and harmless.

ywwg commented:
Daniel, I have to overrule you on this.  We need to be able to fix the cyclic symlink problem (which seems to be extremely rare) without introducing this new problem (which will affect everyone who has duplicate tracks due to symlinks).  @uklotzde do you have any ideas for how we can approach this differently?

uklotzde commented:
Currently I have no clue how Mixxx detects and handles duplicate files. 
Proper normalization of track locations is another item on my wish or task 
list. This will also affect the way how to detect duplicate files.

On 09/16/2015 11:04 PM, Owen Williams wrote:
>
> Daniel, I have to overrule you on this. We need to be able to fix the 
> cyclic symlink problem (which seems to be extremely rare) without 
> introducing this new problem (which will affect everyone who has duplicate 
> tracks due to symlinks). @uklotzde <https://github.com/uklotzde> do you 
> have any ideas for how we can approach this differently?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/708#issuecomment-140893888>.
>

daschuer commented:
@ywww: I don't think we have a procedure of overruling someone in our team. 
We should always head to a solution that fits to a majority. 

For me this PR is a solution for a blocker bug since it fixes a Mixxx crash on machine with link loops. It also fixes the issue with duplicated Tracks. After applying this patch, a new user is not able to get into the weird duplicate condition we discussed above. 
The swapping issue is an issue, but more a small annoyance, compared to a Mixxx crash, which can be fixed manually.  
If we consider this solution it is not good enough, we have another major 1.12 blocker, which we really don't need. 

What do others think?

ywwg commented:
This PR trades one very rare edge case for a new set of bugs that are unknown in prevalence.  Now is not the time to change how mixxx handles duplicate tracks via symlinks.  

With my release manager hat I need to be able to make decisions to keep the project on track for release.  Part of that is making tough decisions about what bugs are worth fixing and which ones can be allowed to persist.  I don't think this fix is worth the problems it introduces.

An acceptable bandaid solution would be to only follow the first symlink, and none following.  Can we do that?

daschuer commented:
> An acceptable bandaid solution would be to only follow the first symlink, and none following. Can we do that?

Unfortunately this will introduce issues that last longer. 

* With a single link solution a undefined number of tracks are marked missing and we may have still duplicates. The user is not able to fix the issue. 

* This PR has no duplicated tracks, which 1.11 and 1.12 can't handle nicely anyway. 
We also have no use case or feature request that includes duplicated tracks. 
A remaining issue is that the location may jump in special conditions. The user is able to fix the duplicated issue manually or later once we have a solution. 

For my feeling all the discussed annoyances from this PR will unlikely really happen. 
Are we hunting a phantom?  

By the way, this PR is also on the track of the fix of this related 1.12 bug: 
https://bugs.launchpad.net/mixxx/+bug/1451778

ywwg commented:
> no use case or feature request that includes duplicated tracks.

But we do have users who are experiencing duplicated tracks, and this PR makes that worse because now their metadata will flip-flop between the two states.  That's worse in my mind than having duplicated tracks.

Since cyclic symlinks are rare, I'd prefer to do nothing and solve the symlink / duplicated track issue more thoroughly in trunk.  Or maybe just try to detect the cyclic issue and throw up a message saying "you have symlink cycles, don't do that.  Scan failed."  That's better than crashing anyway.

daschuer commented:
We can solve the flip-flop issue, by scanning the known directories listed in DB first.
Will you accept such a solution?

ywwg commented:
OK so with that solution:

* users scanning for the first time will see only one of their dupes
* users with dupes will have one of the dupes chosen at random, and the other will be missing
* on subsequent rescans, users with dupes should always see the same track.

Oh and we already have a view for missing tracks, so if the user wanted metadata from the other version, they can do their own merge.

Yes I think that solution would work!

daschuer commented:
Also related: https://bugs.launchpad.net/mixxx/+bug/1497153

daschuer committed:
Exclude and track unhashed directories, include root directories in duplicate check

daschuer committed:
Introduce a second stage looking for unhashed directories

daschuer committed:
Cache security token as well

daschuer committed:
mark not verified tracks under the library tree as deleted

daschuer committed:
Track added Tracks, improve comments in detectMovedFiles

daschuer committed:
pipe added Tracks to detectMovedTracks

daschuer committed:
Added bypass if we have no added tracks

daschuer commented:
The swapping issue in this PR was actually a little bit different than explained above. When it happens the path found in the previous scan was not marked as deleted, because the track is still reachable under both paths. 

A solution, working like discussed was proposed here https://github.com/uklotzde/mixxx/pull/2

Unfortunately, it was more than I had expected, since I wanted to fix the case sensitivity issue in bug https://bugs.launchpad.net/mixxx/+bug/1497153 as well and I had tricked myself during testing this original PR. 

A remaining issue for upgrades and users suffering https://bugs.launchpad.net/mixxx/+bug/1497153
Is, that there is no merging feature in the missing track view.

I think we need at least a feature, that allows to purge a missing track and redirect all its references in playlists and crates to an existing track and increase its play counter accordingly. 

How about a message box, proposing these actions after clicking the purge button:
"Track >artist - title< is referenced in playlist and crates, should they be redirected to similar track 
>....>"?

Other option "Retarget" button 

Ideas?

uklotzde commented:
Unfortunately I experience spurious crashes with the extensions proposed by @daschuer:
https://github.com/uklotzde/mixxx/pull/2

daschuer committed:
Fix two race conditions caused by wrong signal slot connections

daschuer committed:
Fix parameter order of ImportFilesTask, broken since fdb844158be230ffffd0c671d2682961cc975e3a

daschuer committed:
Fix an other race, with tracking success of scanner threads

uklotzde committed:
Merge pull request #2 from daschuer/CyclicSymlinkWorkaround

I haven't checked all changes and all corner cases of moved/deleted files. At least scanning seems to be stable now.

uklotzde committed:
Fix typos

daschuer commented:
@uklotzde: Thank you! Do you have any test results?

uklotzde commented:
I have no test results for moved/deleted files.

Is it be possible to create a TAR archive with a directory structure that contains symbolic links and some tiny audio files just for testing purposes? It would make sense to have the same test data available for all participants. We could place this file in the repository together with the audio files for testing purposes.

daschuer commented:
This is possible. I will create one. However the main benefit for such a file is the later use against regressions.

How about the segfaults? Are yours gone? 
I have fixed some serious race conditions, so we should make sure they will be part of 1.12.

daschuer committed:
Added tar archive with a reference library structure that should now work without any duplicated tracks

uklotzde committed:
Merge pull request #3 from daschuer/CyclicSymlinkWorkaround

Added tar archive with a reference library structure that should now …

uklotzde committed:
Revert "Added tar archive with a reference library structure that should now …"

uklotzde committed:
Merge pull request #4 from uklotzde/revert-3-CyclicSymlinkWorkaround

Revert "Added tar archive with a reference library structure that sho…

uklotzde committed:
Add tar archive with sample library structure including symlinks

Added tar archive with a reference library structure that should
now work without any duplicated tracks

ywwg commented:
symlinks don't work on windows, so the test may have to be linux and/or mac-specific.  That's ok, we'll just ifdef the tests on windows.

daschuer commented:
Is this PR merge-able now?

uklotzde commented:
I will do a final test scanning a directory with around 3500 files and some 
recursive symlinks.

On 10/08/2015 08:14 AM, Daniel Schürmann wrote:
>
> Is this PR merge-able now?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/708#issuecomment-146430414>.
>

uklotzde commented:
Everything looked ok until I finally tried to rename a directory that was 
itself a symlink:

Debug [LibraryScanner 1]: TrackDAO::detectMovedTracks looking for a 
successor of ...
Debug [LibraryScanner 1]: src/library/dao/trackdao.cpp 1698 FAILED QUERY [ 
"SELECT track_locations.id FROM track_locations INNER JOIN library ON 
track_locations.id=library.location WHERE location IN (...) AND filename=? 
AND duration=?" ] QSqlError(-1, "Parameter count mismatch", "")

The query has been modified by the following commit:
939fa4299963f0af88d5486922f70bd757ad42c9 'Added bypass if we have no added 
tracks'

On 10/08/2015 08:35 AM, Uwe Klotz wrote:
> I will do a final test scanning a directory with around 3500 files and 
> some recursive symlinks.
>
> On 10/08/2015 08:14 AM, Daniel Schürmann wrote:
>>
>> Is this PR merge-able now?
>>
>> —
>> Reply to this email directly or view it on GitHub 
>> <https://github.com/mixxxdj/mixxx/pull/708#issuecomment-146430414>.
>>
>

uklotzde committed:
Fix ambiguous SQL query when relocating moved files

A query failed with 'Parameter count mismatch' after renaming a directory
and rescanning the library.

uklotzde commented:
After fixing the ambiguous DB query everything works as expected.

uklotzde commented:
LGTM. Both manual and automatic tests have passed.

ywwg commented:
what test accesses the tar file that was added in this PR?

uklotzde commented:
Unfortunately we currently don't have any unit tests for recursive directory 
structures. The archive is unused.

Should we delete it?

On 10/08/2015 03:15 PM, Owen Williams wrote:
>
> what test accesses the tar file that was added in this PR?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/708#issuecomment-146541422>.
>

daschuer commented:
I would like to keep it for reference during manual test.

daschuer commented:
Thank you for your work and review!

ywwg commented:
please add a readme file describing the file and how it might be used.  Doesn't have to be as detailed as a HOWTO, but just a hint

ywwg commented:
> We can solve the flip-flop issue, by scanning the known directories listed in DB first.

did this change happen?

daschuer commented:
Yes.
